### PR TITLE
[Calyx] Replace calyx.program with module attribute calyx.entrypoint

### PR DESCRIPTION
**Pull request by andrewb1999**
Based on discussion about removing calyx.program in #3565. Allows the calyx dialect to integrate better with other dialects.

### CODE REVIEW

**Comment by andrewb1999** [comment]
CC: @mortbopet @mikeurbach @cgyurgyik @rachitnigam

**Review by cgyurgyik — APPROVED** [review]
Thanks for undertaking this! Approved when the continuous integrations pass.

**Comment by andrewb1999** [comment]
@cgyurgyik Looks like this passed CI, good to merge?

**Comment by cgyurgyik** [comment]
Yes, would you like to merge it yourself?

If so, you'll need to request commit access (below). If not, I can merge it for you.

https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access

File: include/circt/Dialect/Calyx/CalyxLoweringUtils.h
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxLoweringUtils.h:526]: stale comment, function name

File: include/circt/Dialect/Calyx/CalyxLoweringUtils.h
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxLoweringUtils.h:540]: stale variable name

File: include/circt/Dialect/Calyx/CalyxLoweringUtils.h
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxLoweringUtils.h:608]: stale variable name

File: include/circt/Dialect/Calyx/CalyxLoweringUtils.h
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxLoweringUtils.h:618]: stale variable name

File: include/circt/Dialect/Calyx/CalyxLoweringUtils.h
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxLoweringUtils.h:625]: stale variable name

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:1222]: stale variable name

File: lib/Dialect/Calyx/CalyxOps.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/CalyxOps.cpp:1178]: stale variable name

File: lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp:366]: is this needed?

File: lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp
Comment (mortbopet) [inline lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp:402]: stale function name

File: test/Conversion/CalyxToFSM/lower-seq.mlir
Comment (mortbopet) [inline test/Conversion/CalyxToFSM/lower-seq.mlir:34]: These `calyx.entrypoint` attributes shouldn't be necessary for the `CalyxToFSM` passes to succeed.

**Review by mortbopet — COMMENTED** [review]
Some nits.

File: test/Conversion/CalyxToFSM/lower-seq.mlir
Comment (andrewb1999) [reply test/Conversion/CalyxToFSM/lower-seq.mlir:34]: Yeah it's not required as far as I know. Should I leave them out of the test cases?

File: lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp
Comment (andrewb1999) [reply lib/Dialect/Calyx/Transforms/CalyxLoweringUtils.cpp]: No, removed.

**Comment by andrewb1999** [comment]
@mortbopet look good now?

File: test/Conversion/CalyxToFSM/lower-seq.mlir
Comment (mortbopet) [reply test/Conversion/CalyxToFSM/lower-seq.mlir:34]: I'd do that - tests should be as stripped as possible.

**Review by mortbopet — APPROVED** [review]
LGTM - Thanks!

One thing that we should consider for follow-up work is if we even _want_ a `calyx.entrypoint` attribute - I do not think that this should be an inherent property of the IR. As far as I can tell, that entry point only really makes sense for simulators, and not for the compiler itself.  
IMO, tools that require an explicit entry point should:
1. topologically sort the instance hierarchy to infer the top-level component _or_
2. explicitly ask the user about the entry point

A path to this is to implement `HWModuleLike, HWInstanceLike` interfaces for the Calyx ops, and use the [InstanceGraphBase.h](https://github.com/llvm/circt/blob/main/include/circt/Dialect/HW/InstanceGraphBase.h).
(looking at the code, I think that that graph should contain a `getInferredTopModule` operation which performs the topological sort...).

**Pull request merged**
